Add docs regaring --to-remote option for add/import-url (#2091)
iterative/dvc.org: content/docs/command-reference/add.md

```usage
usage: dvc add [-h] [-q | -v] [-R] [--no-commit] [--external]
               [--file <filename>] [--glob] [-o <path>]
               [--to-remote] [-r <name>] [-j <number>]

>                [--desc <text>]

Reviewer: Were the flags actually reordered in core? Otherwise let's please keep the original order, just wrap as needed indeed (under 72 chars).

Author: These were already ordered, though when I checked with core I saw `--glob` came before `--file` so I've also fixed that

>                [--to-remote] [-r <name>] [-j <number>]
>                [--desc <text>]

Reviewer: `-r` will be merged into `--to-remote` according to https://github.com/iterative/dvc/pull/5198#discussion_r556712627, so:
```suggestion
               [--to-remote] [-j <number>] [--desc <text>]
```

Reviewer 2: Those are separate right now, so let's keep them separate in the docs for now.

Reviewer 2: `-r name` also makes sense in a feature of being able to specify default remote for a specific data, so it makes sense to keep `-r name` here.

Reviewer: OK. We did agree to merge them right? In https://github.com/iterative/dvc/pull/5198#pullrequestreview-566902933

Is there an issue to followup on those details? Thanks

p.s. the ideas is that `--to-remote` can accept the remote name as (optional) argument @efiop

> intermediate and final results (like ML models).

After checking that each `target` hasn't been added before (or tracked with
other DVC commands), a few actions are taken under the hood (if `--to-remote` is

> not provided):

Reviewer: I think let's revert this change and be a bit more specific by mentioning to-remote in steps 2 and 3 instead.

other DVC commands), a few actions are taken under the hood:

1. Calculate the file hash.

> 2. Move the file contents to the cache (by default in `.dvc/cache`) (if
>    `--to-remote` option given, then move them to the remote storage), using the
>    file hash to form the cached file path. (See

Reviewer: Copy edit
```suggestion
2. Move the file contents to the cache (by default in `.dvc/cache`) (or to
   remote storage if `--to-remote` is given), using the file hash to form the
   cached file path. (See
```

   [Structure of cache directory](/doc/user-guide/dvc-internals#structure-of-the-cache-directory)
   for more details.)

> 3. Attempt to replace the file with a link to the cached data (unless
>    `--to-remote` option given) (more details on file linking further down).

Reviewer: Copy edit
```suggestion
3. Attempt to replace the file with a link to the cached data (more details on
   file linking further down). Skipped if `--to-remote` is used.
```

`reflink` support, but they have to be specified manually. Refer to the
`cache.type` config option in `dvc config cache` for more information.

### Transferring data directly to the remote

> Giving `--to-remote` option would change the behavior described above. Instead

Reviewer 3: Would be great to start with some motivation, a problem it solves. Like - we'd like to add data from a cloud, but with regular DVC flow it would require downloading the whole dataset first to the machine first - It can be time consuming, we might not have enough space, etc. Let's try to describe a problem first to the best of knowledge, we'll help to iterate on it.

of only being able to give it something from local/remote workspace, it would be
able to support all kinds of remote locations that you can import something
(listed in [import-url](/doc/command-reference/import-url)). The main difference
is that it won't actually do anything on the workspace beside creating a DVC

> file. It will take the data in batches from the given target and transfer it

Reviewer 3: I think `in batches` and `through the local system` are implementation details. We can make them a note if they are important details for the end user. But otherwise it's better to keep this a bit higher level and omit such details.

`reflink` support, but they have to be specified manually. Refer to the
`cache.type` config option in `dvc config cache` for more information.

### Transferring data directly to the remote

> 

Reviewer: I think this new section should go after **Adding entire directories**.

Reviewer: Also, how about

```suggestion
### Transferring data directly to remote storage

```

> When you have a very big dataset that you want to move from some remote location
> to one of your remotes, but at the same time you don't have time or resources to
> store it locally on your local system, you can use `--to-remote` to add that
> remote location straight to remote instead of your local workspace. The remote
> location can be any of the ones that are listed under
> [import-url](/doc/command-reference/import-url) page. When you add a remote
> location with `--to-remote`, it will get the dataset from the given location and
> transfer it to the remote you specified (or the default one). It will create a
> DVC file just like you added something locally, but there won't be any data that
> you can access, unless you [pull](/doc/command-reference/pull) it. In that case,
> it will pull it from the remote storage unit to your workspace and you can start
> using it.

Reviewer: Copy edit:
```suggestion
When you have a very big dataset that you want to move from some external location
to [remote storage](/doc/command-reference/remote) while avoiding storing it locally,
you can use the `--to-remote` option. This will transfer the target data
to a remote of your choice (or the default one). A `.dvc` file will be created normally,
but the data won't be found in your local project until you `dvc pull` it.
```
> Needs formatting.

> This flag is extremely useful when your current system can't handle the data as
> whole, but you still want to track and store it in a remote storage unit, so
> that whenever you switch to a different system that can handle it as a whole (or
> partially) you can simply get the data and start working on it.

Reviewer: Copy edit but I'm not sure I get this motivation, it's kind of implied in the previous paragraph? Maybe not needed

```suggestion
This option is useful when the local system can't handle the target data, but
you still want to track and store it in remote storage, so that whenever you
switch to a different system that can handle it, you can simply pull the data
and start working on it.
```

  > Note that external outputs typically require an external cache setup. See
  > link above for more details.

> - `--to-remote` - Import an external target, but don't move it into the

Reviewer 3: 💅 Import -> import (and in other places I think)

Reviewer: ```suggestion
- `--to-remote` - import an external target, but don't move it into the
```

  > Note that external outputs typically require an external cache setup. See
  > link above for more details.

> - `--to-remote` - transfer data straight to remote, when the used system doesn't

Reviewer 3: same here, a bit too much with the details. we can be brief here, and put a backlink to the section above since we already have it.

I would start with something a bit more clear in a sense that we still use `dvc add` terminology. 

`Add target data into DVC and create a .dvc file, but instead of caching it into DVC cache, transfer it straight to remote storage ... Check this section for the details . If this option is specified target can be any cloud or local URL, not necessarily a local file or directory from the workspace as it is required in the regular `dvc add` workflow. `

  have the means to store it locally. So instead of transferring it to the local
  cache and link it to the working directory, it is transferred through the
  local computer in batches to the remote storage (can be configured using
  `--remote <name>`) and can be checked out locally when the necessary means
  have been established since this process also results with a DVC file.

* `-o <filename>`, `--out <filename>` - destination path for the transferred
  data. (Can only be used with `--to-remote`)

* `-r <name>`, `--remote <name>` - name of the
  [remote storage](/doc/command-reference/remote). (Can only be used with
  `--to-remote`)

* `--desc <text>` - user description of the data (optional). This doesn't affect
  any DVC operations.

> * `-h`, `--help` - prints the usage/help message, and exit.

Reviewer 3: not related? also, pre-commit should be enforcing `-` style

Reviewer: Probably @isidentical's IDE is doing some auto-formatting, please check out https://dvc.org/doc/user-guide/contributing/docs 🙂

Author: hmm weird, perhaps I use the wrong thing somewhere and the autoformatter started replacing others that are below my misusage. should be fixed now!

  > Note that external outputs typically require an external cache setup. See
  > link above for more details.

- `--to-remote` - Track an external target, but don't move it into the
  workspace, nor cache it. [Transfer](#transferring-data-directly-to-the-remote)

>   it directly to remote storage instead. Use `dvc pull` to get the data locally.

Reviewer: > Track an external target

So does this option have any effect or throw an error if `--external` isn't used @isidentical ? Maybe we need an example after all.

Reviewer: Yeah I just tried this and `--external` seems to be needed. Should be mentioned here?

Author: What do you mean by `--external` is needed?

Reviewer: I tried `dvc add --to-remote /external/path` (a default remote exists) and it failed. You have to combine `--external` too right? Unless I did something wrong. If I'm right, let's mention that requirement here ^ and maybe we should consider automatically applying --external in --to-remote so there's no need to type both?

  > Note that external outputs typically require an external cache setup. See
  > link above for more details.

- `--to-remote` - track a single external target file or directory (with a
  `.dvc` file), but instead of caching and linking it locally,
  [transfer](#transferring-data-directly-to-the-remote) it straight to remote

>   storage.

Author: > Or we can extract that to a separate issue.

I'll add it to the #2121

- `-r <name>`, `--remote <name>` - name of the
  [remote storage](/doc/command-reference/remote) to transfer external target to

>   (can only be used with `--to-remote`).

Reviewer: Merge to previous option.

> - `-o <path>`, `--out <path>` - destination path for the transferred data (can

Reviewer: ```suggestion
- `-o <path>`, `--out <path>` - destination `path` for the transferred data (can
```

  > Note that external outputs typically require an external cache setup. See
  > link above for more details.

> - `--to-remote` - add target data into DVC and create a .dvc file, but instead
>   of caching it into DVC cache, transfer it straight to remote storage. Check
>   [this](#transferring-data-directly-to-the-remote) section for the details. If
>   this option is specified target can be any cloud or local URL, not necessarily
>   a local file or directory from the workspace as it is required in the regular
>   dvc addworkflow.

Reviewer: ```suggestion
- `--to-remote` - track a single external target file or directory (with a `.dvc` file),
  but instead of caching and linking it locally,
  [transfer](#transferring-data-directly-to-the-remote) it straight to remote
  storage.
```

Reviewer: Should we update https://dvc.org/doc/user-guide/managing-external-data too and link from here, like it's linked from `--external`?

> - `-o <path>`, `--out <path>` - destination path for the transferred data. (Can
>   only be used with `--to-remote`)
> 
> - `-r <name>`, `--remote <name>` - name of the
>   [remote storage](/doc/command-reference/remote). (Can only be used with
>   `--to-remote`)

Reviewer: ```suggestion
- `-r <name>`, `--remote <name>` - name of the
  [remote storage](/doc/command-reference/remote) to transfer external target to
  (can only be used with `--to-remote`).

- `-o <path>`, `--out <path>` - destination path for the transferred data (can
  only be used with `--to-remote`).
```

But the need for `--out` is unclear to me, please see https://github.com/iterative/dvc/pull/5198#pullrequestreview-569646959
